### PR TITLE
Allow AsyncAPI mocking (WS & Kafka)

### PR DESCRIPTION
**Pull request by JulienBreux**
### Description

- WS
- Kafka

### Related issue(s)

- https://github.com/microcks/microcks-testcontainers-go/issues/8
- https://github.com/microcks/microcks-testcontainers-go/issues/9

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -4,42 +4,43 @@ go 1.21.3
 
 require (
 	github.com/stretchr/testify v1.9.0
-	github.com/testcontainers/testcontainers-go v0.30.0
+	github.com/testcontainers/testcontainers-go v0.31.1-0.20240524121510-d4a21ea92ee8
```
Comment (JulienBreux) [inline go.mod:7]: ⚠️ We're waiting for v0.31.0, blocked by release of fix https://github.com/testcontainers/testcontainers-go/pull/2551

**Comment by lbroudoux** [comment]
Related issue for Go: #8

File: ensemble/async/connection/kafka/kafka.go [hunk 2]
```diff
@@ -13,17 +13,10 @@
...
-// AmazonServiceConnection represents an Amazon service connection settings.
-type AmazonServiceConnection struct {
-	region           string
-	endpointOverride string
-	accessKey        string
-	secretKey        string
+// Connection represents broker connection settings.
+type Connection struct {
+	// BootstrapServers represents the list of bootstrap servers.
+	BootstrapServers string
```
Comment (lbroudoux) [inline ensemble/async/connection/kafka/kafka.go:21]: Could you explain the rationale for splitting the file into two and putting those in dedicated packages? Is it a matter of naming? Of Go conventions or best practices?

File: microcks.go [hunk 3]
```diff
@@ -91,44 +97,76 @@ func WithSecondaryArtifact(artifactFilePath string) testcontainers.CustomizeRequ
...
 		}
 		req.LifecycleHooks = append(req.LifecycleHooks, hooks)
+
+		return nil
 	}
 }
 
-// WithNetwork allows to add a custom network
+// WithNetwork allows to add a custom network.
+// Deprecated: Use network.WithNetwork from testcontainers instead.
```
Comment (lbroudoux) [inline microcks.go:113]: Shouldn't we track this with a new issue? (just after this PR is merged)

File: microcks.go [hunk 4]
```diff
@@ -91,44 +97,76 @@ func WithSecondaryArtifact(artifactFilePath string) testcontainers.CustomizeRequ
...
+func WithSecret(s client.Secret) testcontainers.CustomizeRequestOption {
+	return func(req *testcontainers.GenericContainerRequest) error {
+		hooks := testcontainers.ContainerLifecycleHooks{
+			PostReadies: []testcontainers.ContainerHook{
+				createSecretHook(s),
+			},
+		}
+		req.LifecycleHooks = append(req.LifecycleHooks, hooks)
+
+		return nil
```
Comment (lbroudoux) [inline microcks.go:165]: So this PR also implements #9! Nice side effect 😉

File: microcks.go [hunk 5]
```diff
@@ -91,44 +97,76 @@ func WithSecondaryArtifact(artifactFilePath string) testcontainers.CustomizeRequ
...
+func WithHostAccessPorts(hostAccessPorts []int) testcontainers.CustomizeRequestOption {
+	return func(req *testcontainers.GenericContainerRequest) error {
+		req.HostAccessPorts = hostAccessPorts
+
+		return nil
+	}
+}
+
+// WithSecret allows to add a new secret.
+func WithSecret(s client.Secret) testcontainers.CustomizeRequestOption {
```
Comment (lbroudoux) [inline microcks.go:156]: We're using the `client.Secret` interface here. Shouldn't we remove the local interface type definition above?

File: ensemble/async/connection/kafka/kafka.go (hunk 2 above)
Comment (JulienBreux) [reply ensemble/async/connection/kafka/kafka.go:21]: Yes in fact, this is the way in go to have "namespace" and separated concerns.

**Review by lbroudoux — CHANGES_REQUESTED** [review]
Excellent! Thanks a lot! I just have one concern regarding the definition of an interface type `Secret` in addition of the `client.Secret` that seems to be the one actually used. Can you check this?

File: microcks.go (hunk 5 above)
Comment (JulienBreux) [reply microcks.go:150]: Clearly, done ✅

File: microcks.go (hunk 3 above)
Comment (JulienBreux) [reply microcks.go:107]: Yes, in fact!

**Comment by JulienBreux** [comment]
@lbroudoux juste done about secret!

**Review by lbroudoux — APPROVED** [review]
LGTM!

**Pull request merged**